test: migrated test/message to JS fixture tests

projectnoa opened the pull request:
test: migrated tests/message to JS fixture tests

Migrated all test located in the test/message directory to the fixtures structure and verified the output to fit the .snapshot representation. All tests passed, except for the following, which were not touched. Documentation has been updated to not reference test/message as an example to run test and has been changed to test/report.

Failing tests:
out/Release/node --expose-internals /home/upx/Documents/Development/Javascript/node/test/parallel/test-fs-copyfile.js
out/Release/node /home/upx/Documents/Development/Javascript/node/test/parallel/test-fs-cp.mjs

Fixes: https://github.com/nodejs/node/issues/47707
Refs: https://eslint.org/docs/rules/space-in-parens.html

MoLow commented on deps/zlib/zconf.h.included:
```diff
@@ -0,0 +1,562 @@
+/* zconf.h -- configuration of the zlib compression library
```
what is this dependency?

projectnoa replied:
@MoLow For some reason, my VSCode generated the file automatically when building MAKE. Will remove it.

projectnoa replied:
File has been removed and the original has been restored.

MoLow reviewed (changes requested):
1. please remove the changed zlib file inside dpes
2. nothing requires these added fixtures. a test file must execute them using `common.assertSnapshot` - see https://github.com/nodejs/node/blob/bd913bee367e1ecd308bb7a3e5f6869813add31e/test/parallel/test-node-output-console.mjs and the other tests added in https://github.com/nodejs/node/pull/47498

projectnoa commented:
Hi @MoLow,

As requested, I reverted the changes to the zlib files and added the test runner files in the /parallel directory.

- test-node-output-assertion.mjs
- test-node-output-map.mjs
- test-node-output-promise.mjs
- test-node-output-tick.mjs

Additionally, I modified the test-node-output-console.mjs and test-node-output-errors.mjs to run some of the tests there. Of course, I made sure that the .snapshot files are congruent with the test output and all test succeed.

I was, however, unable to figure out how to make the test util-inspect-error-cause.js evaluate correctly with the snapshot since the output seems to have some special characters that don't play well with RegEx.

Here's an example:

`My attempts to turn the following:

        at Object.<anonymous> \x1B[90m(/\x1B[39mtest*util-inspect-error-cause.js:*:*\x1B[90m)\x1B[39m
    \x1B[90m    at Module._compile (node:internal*modules*cjs*loader:*:*)\x1B[39m
    \x1B[90m    at Module._extensions..js (node:internal*modules*cjs*loader:*:*)\x1B[39m
    \x1B[90m    at Module.load (node:internal*modules*cjs*loader:*:*)\x1B[39m
    \x1B[90m    at Module._load (node:internal*modules*cjs*loader:*:*)\x1B[39m
    \x1B[90m    at Function.executeUserEntryPoint [as runMain] (node:internal*modules*run_main:*:*)\x1B[39m
    \x1B[90m   at node:internal*main*run_main_module:*:*\x1B[39m

into:

        at *
    *[90m    at *[39m
    *[90m    at *[39m
    *[90m    at *[39m
    *[90m    at *[39m
    *[90m    at *[39m
    *[90m    at *[39m
 
where not successful.`

This is the RegEx I was using:

`
const regex = /(\x1B\[\d+m|\(.*\)|<.*>|\/|\w+:)(?=[*\n])/g;
const outputString = inputString.replace(regex, '*');

console.log(outputString);
`

I would appreciate the input and help.

Thanks!

![Screenshot 2023-05-05 at 5 09 00 PM](https://user-images.githubusercontent.com/102206/236588323-db550922-123b-4448-9e96-91dd6790ae56.png)

projectnoa commented on test/parallel/test-node-output-errors.mjs:
```diff
@@ -43,6 +47,16 @@ describe('errors output', { concurrency: true }, () => {
+    { name: 'errors/timeout_throw.js', transform: errTransform },
+    { name: 'errors/undefined_reference_in_new_context.js', transform: errTransform },
+    { name: 'errors/util_inspect_error.js', transform: customTransform },
+    { name: 'errors/util-inspect-error-cause.js', transform: customTransform },
```
@MoLow 

This test case is the one causing the headaches.

See: https://github.com/nodejs/node/pull/47868#issuecomment-1536937028

projectnoa replied:
Resolved by using { tty: true } and using a custom normalization

MoLow commented on test/parallel/test-node-output-tick.mjs:
```diff
@@ -0,0 +1,36 @@
+  ];
+  for (const { name, transform } of tests) {
+    it(name, async () => {
+      await snapshot.spawnAndAssert(fixtures.path(name), transform ?? defaultTransform);
```
```suggestion
      await snapshot.spawnAndAssert(fixtures.path(name), transform ?? defaultTransform, { tty: true });
```

MoLow commented on test/parallel/test-node-output-promise.mjs:
```diff
@@ -0,0 +1,35 @@
+  ];
+  for (const { name, transform } of tests) {
+    it(name, async () => {
+      await snapshot.spawnAndAssert(fixtures.path(name), transform ?? defaultTransform);
```
```suggestion
      await snapshot.spawnAndAssert(fixtures.path(name), transform ?? defaultTransform, { tty: true });
```

MoLow commented on test/parallel/test-node-output-map.mjs:
```diff
@@ -0,0 +1,46 @@
+  ];
+  for (const { name, transform } of tests) {
+    it(name, async () => {
+      await snapshot.spawnAndAssert(fixtures.path(name), transform ?? defaultTransform);
```
```suggestion
      await snapshot.spawnAndAssert(fixtures.path(name), transform ?? defaultTransform, { tty: true });
```

MoLow commented on test/parallel/test-node-output-errors.mjs:
```diff
@@ -43,6 +47,16 @@ describe('errors output', { concurrency: true }, () => {
     { name: 'errors/throw_in_line_with_tabs.js', transform: errTransform },
     { name: 'errors/throw_non_error.js', transform: errTransform },
     { name: 'errors/promise_always_throw_unhandled.js', transform: promiseTransform },
+    { name: 'errors/if-error-has-good-stack.js', transform: errTransform },
```
please add tty: true to the migrated test cases

MoLow commented on test/parallel/test-node-output-assertion.mjs:
```diff
@@ -0,0 +1,36 @@
+  ];
+  for (const { name, transform } of tests) {
+    it(name, async () => {
+      await snapshot.spawnAndAssert(fixtures.path(name), transform ?? defaultTransform);
```
```suggestion
      await snapshot.spawnAndAssert(fixtures.path(name), transform ?? defaultTransform, { tty: true });
```

MoLow commented on test/fixtures/console/core_line_numbers.snapshot:
```diff
@@ -0,0 +1,10 @@
+	throw new RangeError(errors[type]);
+	^
+
+RangeError: Invalid input
```
lets avoid this PR from containing snapshots with a delta larger than the git rename threshold

projectnoa replied:
Apologies for my ignorance @MoLow but I don't know what this means. Could you please explain what you mean? Thanks!

MoLow replied:
if the diff (either via GitHub UI, or via git diff) detects a file as new instead as renamed, I think we should exclude it from this PR, to further PRs, so reviewing would be easier

projectnoa replied:
I see. Should I make a correction and break down the PR? or is this a request for future PR's?

Thanks!

MoLow replied:
it seems like the diff in this screenshot is a diff for a specific commit, not for the entire PR

targos commented:
I don't know what should be the prefix of the test file names, but `test-node-` doesn't seem right.

projectnoa commented:
> I don't know what should be the prefix of the test file names, but `test-node-` doesn't seem right.

I totally agree, but I couldn't figure out a clear pattern or methodology for naming all these diverse tests. Open to suggestions. Thanks @targos!

projectnoa commented:
@MoLow Test runner name prefix changed to test-output.

Let me know of any other observations.

Thanks!

MoLow commented on test/message/core_line_numbers.out:
please revert all deleted files

MoLow commented:
this PR will need to be squashed / remove all merge commits

projectnoa commented:
> this PR will need to be squashed / remove all merge commits

Apologies @MoLow but I am not versed enough in git to figure out how to squash the commits when there's a merge in between. 

Could anybody help me here?

Appreciate the help and I'm sorry for my ignorance.

Thanks!

MoLow commented:
You would probably want to:

```bash
git fetch upstream main
git rebase upstream/main 
git reset --soft upstream/main
git add .
git commit -m "test: migrate message tests to js fixtures"
```

projectnoa committed:
test: migrate message tests to js fixtures squashed

test: migrate message tests to js fixtures squashed

Add test runners to /parallel dir and fix snapshot

Update test/parallel/test-node-output-tick.mjs

Co-authored-by: Moshe Atlow <moshe@atlow.co.il>

Update test/parallel/test-node-output-promise.mjs

Co-authored-by: Moshe Atlow <moshe@atlow.co.il>

Update test/parallel/test-node-output-assertion.mjs

Co-authored-by: Moshe Atlow <moshe@atlow.co.il>

Update test/parallel/test-node-output-map.mjs

Co-authored-by: Moshe Atlow <moshe@atlow.co.il>

fixed snapshot discrepancy with output-errors.mjs

Corrected test runner name prefix

Removing snapshots that show as 'new' on git diff

Restored deleted files

Deleted files restored

projectnoa commented:
> You would probably want to:
> 
> ```shell
> git fetch upstream main
> git rebase upstream/main 
> git reset --soft upstream/main
> git add .
> git commit -m "test: migrate message tests to js fixtures"
> ```

@MoLow I ran the commands and merged the changes. I don't believe the squashing of the commits was achieved tho. I'm struggling to figure this out. I'm holding back this PR with my lack of experience with git.

projectnoa commented:
@MoLow

Should I discard this PR and create a new one with only the changes in one commit?

MoLow commented:
either that, or squash this PR into a commit with no merge commits, and no conflicts

projectnoa commented:
> either that, or squash this PR into a commit with no merge commits, and no conflicts

Figured out how to properly squash the PR commits. Apologies for the hassle.

MoLow commented on test/fixtures/map/source_map_enclosing_function.js:
```diff
@@ -0,0 +1,7 @@
+// Flags:  --enable-source-maps
```
please revert

MoLow commented on test/fixtures/console/core_line_numbers.js:
```diff
@@ -1,5 +1,5 @@
 'use strict';
-require('../common');
+require('../../common');
```
missing snapshot file?

projectnoa committed:
Fix to tests and missing snapshots

projectnoa commented:
@MoLow let me know of any other observations.

Thanks!

projectnoa commented:
Hi @MoLow,

Could you please provide some guidance?

I really appreciate your help.

Thanks!

MoLow commented:
> Could you please provide some guidance?

Sure, you should go over the diff of this PR and undo any changes to files that are not marked as moved

projectnoa commented:
> > Could you please provide some guidance?
> 
> Sure, you should go over the diff of this PR and undo any changes to files that are not marked as moved
@MoLow 
Won't this break the tests? Some files will not have the changes that are needed for the new tests to work.

MoLow commented:
if it brakes tests, lets revert that as well.
I think this PR should have a minimal diff

projectnoa commented:
> if it brakes tests, lets revert that as well.
> I think this PR should have a minimal diff



So, if I understand correctly. 
1. All .out files that were moved into fixtures as a .snapshot file must have the original content (I must not regenerate or modify these outputs).
2. Any .js files that cannot validate this original output must be reverted and left on the messages directory.
3. All .mjs files that run the tests don't apply to the previous rules since they are new.

Is this correct @MoLow ?

MoLow commented:
> All .out files that were moved into fixtures as a .snapshot file must have the original content (I must not regenerate or modify these outputs).

I don't think having the original content is a must. I only asked the diff to be small enough for git to detect it as a file rename

> Any .js files that cannot validate this original output must be reverted and left on the messages directory.

for this phase, yes - then we can iterate in smaller PR's that can be reviewed more carfully

> All .mjs files that run the tests don't apply to the previous rules since they are new.

correct. they also do not have to be mjs files, that was just done for convenience

projectnoa committed:
Restored snapshots to original .out content

projectnoa commented:
Hello @MoLow, @NiharPhansalkar,

I did my best to restore the .snapshots to their original state so they show as Edited and not New. Additionally, I restored the original Python tests in the message directory. I was unable to make the following test pass (stdin_messages.js, eval_messages.js, core_line_numbers.js, util-inspect-error-cause.js) since the contents of some of the .out files is just too different to the output produced by the tests. In the end I was just bending over backwards to twist the code to validate the test, which defeats the purpose of tests.

If anybody is smart enough to figure where I went wrong and want to fix this PR, please do so. Otherwise, this is where I end my journey with this contribution, for my sanity's sake.

I apologize for not being able to figure this out and make a proper contribution.

Feel free to discard this PR if the effort of fixing it is not worth it.

Thanks

MoLow commented:
I don't think we should make the tests reduce the diff. just ship a first PR where all diffs are minimal, excluding tests that are more complicated. then in follow-up PRS - we can handle those tests with a bigger diff

GeoffreyBooth commented on test/fixtures/map/source_map_throw_set_immediate.js:
```diff
@@ -1,5 +1,5 @@
-require('../../common');
-require('../source-map/uglify-throw');
+require('../common');
+require('../fixtures/source-map/uglify-throw');
```
I think in general we want all files to end with a newline.

GeoffreyBooth commented on test/fixtures/errors/throw_error_with_getter_throw.snapshot:
```diff
@@ -1,8 +1,8 @@
+(Use `* --trace-uncaught ...` to show where the exception was thrown)
 
-Node.js *
+Node.js *
```
Why are all these files not ending with newlines? Aren’t they autogenerated?

GeoffreyBooth commented on test/fixtures/console/eval_messages.snapshot:
```diff
@@ -1,81 +1,81 @@
-    at *
-    at *
-    at *
+    at new Script (node:vm:*:*)
+    at createScript (node:vm:*:*)
+    at Object.runInThisContext (node:vm:*:*)
+    at node:internal/process/execution:*:*
+    at [eval]-wrapper:*:*
+    at runScript (node:internal/process/execution:*:*)
+    at evalScript (node:internal/process/execution:*:*)
+    at node:internal/main/eval_string:*:*
```
This stack trace is a step backward; this test will break if this call path ever changes, which isn’t something that should be breaking a test. Can we eliminate this from the snapshot, perhaps by updating the snapshot generation script if necessary?
